src/runtime/driver/mod.rs

```diff
@@ -91,6 +92,198 @@ impl Driver {
+            Lifecycle::Completed(cqe) => {
+                self.ops.remove(op.index());
+                Poll::Ready(op.take_data().unwrap().complete(cqe))
+            }
+            Lifecycle::CompletionList(..) => {
+                unreachable!("No `more` flag set for SingleCQE")
+            }
+        }
+    }
+
+    pub(crate) fn poll_multishot_op<T>(
+        &mut self,
+        op: &mut Op<T, MultiCQEFuture>,
+        cx: &mut Context<'_>,
+    ) -> Poll<T::Output>
+    where
+        T: Unpin + 'static + Completable + Updateable,
+    {
+        use std::mem;
+
+        let (lifecycle, completions) = self
+            .ops
+            .get_mut(op.index())
+            .expect("invalid internal state");
+
+        match mem::replace(lifecycle, Lifecycle::Submitted) {
+            Lifecycle::Submitted => {
+                *lifecycle = Lifecycle::Waiting(cx.waker().clone());
+                Poll::Pending
+            }
+            Lifecycle::Waiting(waker) if !waker.will_wake(cx.waker()) => {
+                *lifecycle = Lifecycle::Waiting(cx.waker().clone());
+                Poll::Pending
+            }
+            Lifecycle::Waiting(waker) => {
+                *lifecycle = Lifecycle::Waiting(waker);
+                Poll::Pending
+            }
+            Lifecycle::Ignored(..) => unreachable!(),
+            Lifecycle::Completed(cqe) => {
+                // This is possible. We may have previously polled a CompletionList,
+                // and the final CQE registered as Completed
+                self.ops.remove(op.index());
+                Poll::Ready(op.take_data().unwrap().complete(cqe))
+            }
+            Lifecycle::CompletionList(indices) => {
+                let mut data = op.take_data().unwrap();
+                let mut status = Poll::Pending;
+                // Consume the CqeResult list, calling update on the Op on all Cqe's flagged `more`
+                // If the final Cqe is present, clean up and return Poll::Ready
+                for cqe in indices.into_list(completions) {
+                    if cqueue::more(cqe.flags) {
+                        data.update(cqe);
+                    } else {
+                        status = Poll::Ready(cqe);
+                        break;
+                    }
+                }
+                match status {
+                    Poll::Pending => {
+                        // We need more CQE's. Restore the op state
+                        let _ = op.insert_data(data);
+                        *lifecycle = Lifecycle::Waiting(cx.waker().clone());
+                        Poll::Pending
+                    }
+                    Poll::Ready(cqe) => {
+                        self.ops.remove(op.index());
+                        Poll::Ready(data.complete(cqe))
+                    }
+                }
+            }
+        }
+    }
+
+    pub(crate) fn remove_op<T, CqeType>(&mut self, op: &mut Op<T, CqeType>) {
```

Comment:
Should `remove_op` go right after `submit_op` a little above in this file?

Reply:
I don't really care much how these are ordered. I'm fine either way.

Reply:
Done.